File: src-tauri/src/mods.rs
use crate::deadlock::{get_addons_path, get_disabled_path};
use crate::error::AppError;
use crate::types::Mod;
use chrono::{DateTime, Utc};
use std::fs;
use std::path::Path;

/// Parse a VPK filename to extract priority
/// Format: pak##_dir.vpk where ## is 01-99
fn parse_vpk_priority(filename: &str) -> Option<u32> {
    if !filename.starts_with("pak") || !filename.ends_with("_dir.vpk") {
        return None;
    }

    let number_part = &filename[3..5];
    number_part.parse::<u32>().ok()
}

/// Generate a mod ID from the file path
fn generate_mod_id(path: &Path) -> String {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    let mut hasher = DefaultHasher::new();
    path.to_string_lossy().hash(&mut hasher);
    format!("{:x}", hasher.finish())
}

/// Extract a human-readable name from the VPK filename
fn extract_mod_name(filename: &str) -> String {
    // Try to extract a meaningful name from pak##_name_dir.vpk format
    let name = filename
        .trim_end_matches("_dir.vpk")
        .trim_end_matches(".vpk");

    // Remove the pak## prefix if present
    let name = if name.starts_with("pak") && name.len() > 5 {
        let rest = &name[5..]; // Skip "pak##"
        if rest.starts_with('_') {
            &rest[1..] // Skip the underscore
        } else {
            rest
        }
    } else {
        name
    };

    // Convert underscores/dashes to spaces and title case
    name.replace(['_', '-'], " ")
        .split_whitespace()
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                None => String::new(),
                Some(first) => first.to_uppercase().chain(chars).collect(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// Scan for mods in the addons folder
pub fn scan_mods(deadlock_path: &Path) -> Result<Vec<Mod>, AppError> {
    let addons_path = get_addons_path(deadlock_path)?;
    let disabled_path = get_disabled_path(deadlock_path)?;

    let mut mods = Vec::new();

    // Scan enabled mods
    mods.extend(scan_folder(&addons_path, true)?);

    // Scan disabled mods
    if disabled_path.exists() {
        mods.extend(scan_folder(&disabled_path, false)?);
    }

    // Sort by priority
    mods.sort_by_key(|m| m.priority);

    Ok(mods)
}

fn scan_folder(folder: &Path, enabled: bool) -> Result<Vec<Mod>, AppError> {
    let mut mods = Vec::new();

    if !folder.exists() {
        return Ok(mods);
    }

    for entry in fs::read_dir(folder)? {
        let entry = entry?;
        let path = entry.path();

        if !path.is_file() {
            continue;
        }

        let filename = match path.file_name().and_then(|n| n.to_str()) {
            Some(name) => name.to_string(),
            None => continue,
        };

        // Only process VPK files
        if !filename.ends_with("_dir.vpk") && !filename.ends_with(".vpk") {
            continue;
        }

        let metadata = fs::metadata(&path)?;
        let modified: DateTime<Utc> = metadata
            .modified()
            .map(|t| t.into())
            .unwrap_or_else(|_| Utc::now());

        let priority = parse_vpk_priority(&filename).unwrap_or(50);

        mods.push(Mod {
            id: generate_mod_id(&path),
            name: extract_mod_name(&filename),
            file_name: filename,
            path: path.to_string_lossy().to_string(),
            enabled,
            priority,
            size: metadata.len(),
            installed_at: modified,
            description: None,
            thumbnail_url: None,
            game_banana_id: None,
        });
    }

    Ok(mods)
}

/// Enable a mod by moving it from disabled to addons folder
pub fn enable_mod(deadlock_path: &Path, mod_id: &str) -> Result<Mod, AppError> {
    let mods = scan_mods(deadlock_path)?;
    let target_mod = mods
        .iter()
        .find(|m| m.id == mod_id)
        .ok_or_else(|| AppError::ModNotFound(mod_id.to_string()))?;

    if target_mod.enabled {
        return Ok(target_mod.clone());
    }

    let addons_path = get_addons_path(deadlock_path)?;
    let source_path = Path::new(&target_mod.path);
    let dest_path = addons_path.join(&target_mod.file_name);

    fs::rename(source_path, &dest_path)?;

    // Return updated mod
    let mut updated_mod = target_mod.clone();
    updated_mod.enabled = true;
    updated_mod.path = dest_path.to_string_lossy().to_string();

    Ok(updated_mod)
}

/// Disable a mod by moving it to the disabled folder
pub fn disable_mod(deadlock_path: &Path, mod_id: &str) -> Result<Mod, AppError> {
    let mods = scan_mods(deadlock_path)?;
    let target_mod = mods
        .iter()
        .find(|m| m.id == mod_id)
        .ok_or_else(|| AppError::ModNotFound(mod_id.to_string()))?;

    if !target_mod.enabled {
        return Ok(target_mod.clone());
    }

    let disabled_path = get_disabled_path(deadlock_path)?;
    let source_path = Path::new(&target_mod.path);
    let dest_path = disabled_path.join(&target_mod.file_name);

    fs::rename(source_path, &dest_path)?;

    // Return updated mod
    let mut updated_mod = target_mod.clone();
    updated_mod.enabled = false;
    updated_mod.path = dest_path.to_string_lossy().to_string();

    Ok(updated_mod)
}

/// Delete a mod completely
pub fn delete_mod(deadlock_path: &Path, mod_id: &str) -> Result<(), AppError> {
    let mods = scan_mods(deadlock_path)?;
    let target_mod = mods
        .iter()
        .find(|m| m.id == mod_id)
        .ok_or_else(|| AppError::ModNotFound(mod_id.to_string()))?;

    fs::remove_file(&target_mod.path)?;

    // Also remove related VPK files (pak##_000.vpk, pak##_001.vpk, etc.)
    let base_name = target_mod.file_name.trim_end_matches("_dir.vpk");
    let parent = Path::new(&target_mod.path)
        .parent()
        .ok_or_else(|| AppError::ModNotFound(mod_id.to_string()))?;

    for entry in fs::read_dir(parent)? {
        let entry = entry?;
        let filename = entry.file_name();
        let filename_str = filename.to_string_lossy();

        if filename_str.starts_with(base_name) && filename_str.ends_with(".vpk") {
            fs::remove_file(entry.path())?;
        }
    }

    Ok(())
}

/// Set the priority of a mod by renaming it
pub fn set_mod_priority(
    deadlock_path: &Path,
    mod_id: &str,
    new_priority: u32,
) -> Result<Mod, AppError> {
    let mods = scan_mods(deadlock_path)?;
    let target_mod = mods
        .iter()
        .find(|m| m.id == mod_id)
        .ok_or_else(|| AppError::ModNotFound(mod_id.to_string()))?;

    let source_path = Path::new(&target_mod.path);
    let parent = source_path
        .parent()
        .ok_or_else(|| AppError::ModNotFound(mod_id.to_string()))?;

    // Generate new filename with new priority
    let priority_str = format!("{:02}", new_priority.min(99));
    let new_filename = format!("pak{}_dir.vpk", priority_str);
    let dest_path = parent.join(&new_filename);

    // Check if destination already exists
    if dest_path.exists() && dest_path != source_path {
        return Err(AppError::InvalidDeadlockPath(format!(
            "Priority {} is already in use",
            new_priority
        )));
    }

    fs::rename(source_path, &dest_path)?;

    // Return updated mod
    let mut updated_mod = target_mod.clone();
    updated_mod.priority = new_priority;
    updated_mod.file_name = new_filename;
    updated_mod.path = dest_path.to_string_lossy().to_string();

    Ok(updated_mod)
}

File: electron/main/services/mods.ts
import {
    readdirSync,
    statSync,
    renameSync,
    unlinkSync,
    existsSync,
} from 'fs';
import { join, basename } from 'path';
import { createHash } from 'crypto';
import { getAddonsPath, getDisabledPath } from './deadlock';

export interface Mod {
    id: string;
    name: string;
    fileName: string;
    path: string;
    enabled: boolean;
    priority: number;
    size: number;
    installedAt: string;
    description?: string;
    thumbnailUrl?: string;
    gameBananaId?: number;
    gameBananaFileId?: number;
    categoryId?: number;
    categoryName?: string;
    sourceSection?: string;
    nsfw?: boolean;
}

/**
 * Parse VPK filename to extract priority (pak##_dir.vpk format)
 */
function parseVpkPriority(filename: string): number | null {
    if (
        !filename.startsWith('pak') ||
        (!filename.endsWith('_dir.vpk') && !filename.endsWith('.vpk'))
    ) {
        return null;
    }
    const numberPart = filename.slice(3, 5);
    const num = parseInt(numberPart, 10);
    return isNaN(num) ? null : num;
}

/**
 * Generate a mod ID from the file name (hash)
 * Uses fileName instead of full path so ID stays stable when mod moves between folders
 */
function generateModId(fileName: string): string {
    return createHash('md5').update(fileName).digest('hex').slice(0, 16);
}

/**
 * Extract a human-readable name from the VPK filename
 */
function extractModName(filename: string): string {
    // Remove _dir.vpk or .vpk suffix
    let name = filename.replace(/_dir\.vpk$/, '').replace(/\.vpk$/, '');

    // Remove pak## prefix if present
    if (name.startsWith('pak') && name.length > 5) {
        const rest = name.slice(5);
        name = rest.startsWith('_') ? rest.slice(1) : rest;
    }

    // Convert underscores/dashes to spaces and title case
    return name
        .replace(/[_-]/g, ' ')
        .split(/\s+/)
        .filter((word) => word.length > 0)
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
}

/**
 * Scan a folder for VPK mods
 */
function scanFolder(folder: string, enabled: boolean): Mod[] {
    const mods: Mod[] = [];

    if (!existsSync(folder)) {
        return mods;
    }

    const entries = readdirSync(folder);

    for (const entry of entries) {
        const fullPath = join(folder, entry);

        try {
            const stats = statSync(fullPath);
            if (!stats.isFile()) continue;

            // Only process VPK files
            if (!entry.endsWith('_dir.vpk') && !entry.endsWith('.vpk')) continue;

            const priority = parseVpkPriority(entry) ?? 50;

            mods.push({
                id: generateModId(entry),
                name: extractModName(entry),
                fileName: entry,
                path: fullPath,
                enabled,
                priority,
                size: stats.size,
                installedAt: stats.mtime.toISOString(),
            });
        } catch {
            // Skip files we can't read
        }
    }

    return mods;
}

/**
 * Scan for all mods in both enabled and disabled folders
 */
export function scanMods(deadlockPath: string): Mod[] {
    const addonsPath = getAddonsPath(deadlockPath);
    const disabledPath = getDisabledPath(deadlockPath);

    const mods: Mod[] = [
        ...scanFolder(addonsPath, true),
        ...scanFolder(disabledPath, false),
    ];

    // Sort by priority
    mods.sort((a, b) => a.priority - b.priority);

    return mods;
}

/**
 * Find the next available priority number that doesn't conflict
 * Checks BOTH addons and disabled folders to avoid overwriting disabled mods
 */
export function findNextAvailablePriority(deadlockPath: string, startFrom = 1): number {
    const usedPriorities = getUsedPriorities(deadlockPath);

    // Find next available starting from startFrom (default 1)
    let priority = startFrom;
    while (usedPriorities.has(priority) && priority < 99) {
        priority++;
    }

    // If all numbers up to 99 are taken, this is an error
    if (priority >= 99 && usedPriorities.has(99)) {
        throw new Error('No available priority slots (all 1-99 are used)');
    }

    return priority;
}

/**
 * Get the set of used priorities in BOTH addons AND disabled folders
 * This prevents conflicts when downloading new mods
 */
export function getUsedPriorities(deadlockPath: string): Set<number> {
    const addonsPath = getAddonsPath(deadlockPath);
    const disabledPath = getDisabledPath(deadlockPath);
    const usedPriorities = new Set<number>();

    // Check addons folder
    if (existsSync(addonsPath)) {
        const entries = readdirSync(addonsPath);
        for (const entry of entries) {
            const priority = parseVpkPriority(entry);
            if (priority !== null) {
                usedPriorities.add(priority);
            }
        }
    }

    // Also check disabled folder
    if (existsSync(disabledPath)) {
        const entries = readdirSync(disabledPath);
        for (const entry of entries) {
            const priority = parseVpkPriority(entry);
            if (priority !== null) {
                usedPriorities.add(priority);
            }
        }
    }

    return usedPriorities;
}

/**
 * Enable a mod by moving it from disabled to addons folder
 */
export function enableMod(deadlockPath: string, modId: string): Mod {
    const mods = scanMods(deadlockPath);
    const targetMod = mods.find((m) => m.id === modId);

    if (!targetMod) {
        throw new Error(`Mod not found: ${modId}`);
    }

    if (targetMod.enabled) {
        return targetMod;
    }

    const addonsPath = getAddonsPath(deadlockPath);
    const destPath = join(addonsPath, targetMod.fileName);

    renameSync(targetMod.path, destPath);

    return {
        ...targetMod,
        enabled: true,
        path: destPath,
    };
}

/**
 * Disable a mod by moving it to the disabled folder
 */
export function disableMod(deadlockPath: string, modId: string): Mod {
    const mods = scanMods(deadlockPath);
    const targetMod = mods.find((m) => m.id === modId);

    if (!targetMod) {
        throw new Error(`Mod not found: ${modId}`);
    }

    if (!targetMod.enabled) {
        return targetMod;
    }

    const disabledPath = getDisabledPath(deadlockPath);
    const destPath = join(disabledPath, targetMod.fileName);

    renameSync(targetMod.path, destPath);

    return {
        ...targetMod,
        enabled: false,
        path: destPath,
    };
}

/**
 * Delete a mod completely (including related VPK files)
 */
export function deleteMod(deadlockPath: string, modId: string): void {
    const mods = scanMods(deadlockPath);
    const targetMod = mods.find((m) => m.id === modId);

    if (!targetMod) {
        throw new Error(`Mod not found: ${modId}`);
    }

    // Delete the main file
    unlinkSync(targetMod.path);

    // Also remove related VPK files (pak##_000.vpk, pak##_001.vpk, etc.)
    const baseName = targetMod.fileName.replace(/_dir\.vpk$/, '');
    const parentDir = join(targetMod.path, '..');

    try {
        const siblings = readdirSync(parentDir);
        for (const sibling of siblings) {
            if (sibling.startsWith(baseName) && sibling.endsWith('.vpk')) {
                unlinkSync(join(parentDir, sibling));
            }
        }
    } catch {
        // Ignore errors when cleaning up related files
    }
}

/**
 * Set the priority of a mod by renaming it
 */
export function setModPriority(
    deadlockPath: string,
    modId: string,
    newPriority: number
): Mod {
    const mods = scanMods(deadlockPath);
    const targetMod = mods.find((m) => m.id === modId);

    if (!targetMod) {
        throw new Error(`Mod not found: ${modId}`);
    }

    const parentDir = join(targetMod.path, '..');
    const priorityStr = String(Math.min(99, newPriority)).padStart(2, '0');

    // Preserve the mod name by only replacing the priority prefix
    // e.g., pak05_cool_skin_dir.vpk -> pak10_cool_skin_dir.vpk
    const newFileName = targetMod.fileName.replace(/^pak\d{2}_/, `pak${priorityStr}_`);
    const destPath = join(parentDir, newFileName);

    // Check if destination already exists
    if (existsSync(destPath) && destPath !== targetMod.path) {
        throw new Error(`Priority ${newPriority} is already in use`);
    }

    renameSync(targetMod.path, destPath);

    return {
        ...targetMod,
        priority: newPriority,
        fileName: newFileName,
        path: destPath,
        id: generateModId(newFileName),
    };
}
